Cannot unbox tuple

This test case fails with an exception:

```
[case testReadTupleFromList]
from typing import List, Tuple
def f(x: List[Tuple[int, int]]) -> None:
    a = x[0]
[file driver.py]
from native import f
f([(1, 2)])
```

Here's the traceback:

```
/Users/jukka/src/mypyc/mypyc/test/test_run.py:76: in run_case
    alt_lib_path=test_temp_dir)
/Users/jukka/src/mypyc/mypyc/emitmodule.py:53: in compile_modules_to_c
    return generator.generate_c_for_modules()
/Users/jukka/src/mypyc/mypyc/emitmodule.py:113: in generate_c_for_modules
    generate_native_function(fn, emitter, self.source_paths[module_name], module_name)
/Users/jukka/src/mypyc/mypyc/emitfunc.py:61: in generate_native_function
    op.accept(visitor)
/Users/jukka/src/mypyc/mypyc/ops.py:1161: in accept
    return visitor.visit_unbox(self)
/Users/jukka/src/mypyc/mypyc/emitfunc.py:281: in visit_unbox
    self.emitter.emit_unbox(self.reg(op.src), self.reg(op), op.type)
/Users/jukka/src/mypyc/mypyc/emit.py:277: in emit_unbox
    '%s = %s;' % (dest, typ.c_error_value())]
/Users/jukka/src/mypyc/mypyc/ops.py:60: in c_error_value
    return self.c_undefined_value()
/Users/jukka/src/mypyc/mypyc/ops.py:206: in c_undefined_value
    assert False, "Tuple undefined value can't be represented as a C expression"
E   AssertionError: Tuple undefined value can't be represented as a C expression
```

This comes up pretty often if I try to do anything non-trivial with tuples.

--- a/mypyc/emit.py
+++ b/mypyc/emit.py
@@ -26,10 +26,19 @@ class EmitterContext:
         self.temp_counter = 0
         self.names = NameGenerator(module_names)
 
+        # The two maps below are used for generating declarations or
+        # definitions at the top of the C file. The main idea is that they can
+        # be generated at any time during the emit phase.
+
         # A map of a C identifier to whatever the C identifier declares. Currently this is
         # used for declaring structs and the key corresponds to the name of the struct.
         # The declaration contains the body of the struct.
         self.declarations = OrderedDict()  # type: Dict[str, HeaderDeclaration]
+
+        # A map from C identifier to code that defined the C identifier. This
+        # is similar to to 'declarations', but these may appear after the
+        # declarations in the generated code.
+        self.statics = OrderedDict()  # type: Dict[str, str]
 
 
 class Emitter:
@@ -119,11 +128,7 @@ class Emitter:
         elif isinstance(rtype, RPrimitive):
             return rtype.c_undefined
         elif isinstance(rtype, RTuple):
-            # This doesn't work since this is expected to return a C expression, but
-            # defining an undefined tuple requires declaring a temp variable, such as:
-            #
-            #    struct foo _tmp = { <item0-undefined>, <item1-undefined>, ... };
-            assert False, "Tuple undefined value can't be represented as a C expression"
+            return self.tuple_undefined_value(rtype)
         assert False, rtype
 
     def c_error_value(self, rtype: RType) -> str:
@@ -156,6 +161,18 @@ class Emitter:
         result.append('')
 
         return result
+
+    def tuple_undefined_value(self, rtuple: RTuple) -> str:
+        context = self.context
+        id = self.tuple_unique_id(rtuple)
+        name = 'tuple_undefined_' + id
+        if name not in context.statics:
+            struct_name = self.tuple_struct_name(rtuple)
+            values = ', '.join(self.c_undefined_value(item)
+                               for item in rtuple.types)
+            init = 'struct {} {} = {{ {} }};'.format(struct_name, name, values)
+            context.statics[name] = init
+        return name
 
     # Higher-level operations
 

--- a/mypyc/emitmodule.py
+++ b/mypyc/emitmodule.py
@@ -123,6 +123,9 @@ class ModuleGenerator:
         for declaration in self.toposort_declarations():
             declarations.emit_lines(*declaration.body)
 
+        for static_def in self.context.statics.values():
+            declarations.emit_line(static_def)
+
         return ''.join(declarations.fragments + emitter.fragments)
 
     def generate_module_def(self, emitter: Emitter, module_name: str, module: ModuleIR) -> None:

--- a/test-data/run.test
+++ b/test-data/run.test
@@ -949,3 +949,13 @@ from native import from_tuple, from_list, from_any
 assert from_tuple((1, 'x')) == ['x', 1]
 assert from_list([3, 4]) == [4, 3]
 assert from_any('xy') == ['y', 'x']
+
+[case testUnboxTuple]
+from typing import List, Tuple
+
+def f(x: List[Tuple[int, int]]) -> int:
+    a, b = x[0]
+    return a + b
+[file driver.py]
+from native import f
+assert f([(5, 6)]) == 11

